sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -492,12 +492,43 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
  * Remove no-op operators from the query plan that do not make any modifications.
  */
 object RemoveNoopOperators extends Rule[LogicalPlan] {
+  private def removeAliasOnlyProject(plan: LogicalPlan): LogicalPlan = plan match {
+    case p @ Project(projectList, child) =>
+      val originalOutputs = projectList.collect {
+        case Alias(a: Attribute, _) => a
+        case a: Attribute => a
+      }
+      if (child.outputSet.nonEmpty && child.outputSet == AttributeSet(originalOutputs)) {
+        child
+      } else {
+        p
+      }
+    case _ => plan
+  }
+
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     // Eliminate no-op Projects
     case p @ Project(_, child) if child.sameOutput(p) => child
 
     // Eliminate no-op Window
     case w: Window if w.windowExpressions.isEmpty => w.child
+
+    // Eliminate no-op Union
+    case d @ Distinct(u: Union) =>
+      val unionChildren = u.children.map(c => removeAliasOnlyProject(c))
+      if (unionChildren.tail.forall(unionChildren.head.sameResult(_))) {
+        d.withNewChildren(Seq(u.children.head))
+      } else {
+        d
+      }
+
+    case d @ Deduplicate(_, u: Union) =>
```

Comment:
If we handle only two cases (`Distinct` and `Deduplicate`), let's say explicitly `Distinct/Deduplicate` instead of saying `Distinct like operators`. In general, we use a `whitelist` approach.